core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,487 @@
+    // transition and still keep the LB in TRANSIENT_FAILURE state. This is referred to as "sticky
+    // transient failure". Only a subchannel state change to READY will get the LB out of
+    // TRANSIENT_FAILURE. If the state is IDLE we additionally request a new connection so that we
+    // keep retrying for a connection.
+
+    // With the new pick first implementation, iterative requests for connections will not be
+    // requested once the first pass through is complete. This means that individual subchannels
+    // are responsible for coming out of backoff and starting a transport.
+    // However, if the first pass through is complete and we report TRANSIENT_FAILURE but an address
+    // update occurs, the iterative logic will still be present for the first pass through for
+    // the new address list even though we are in a state of TRANSIENT_FAILURE.
+    if (currentState == TRANSIENT_FAILURE) {
+      if (!iterate && (newState == CONNECTING || newState == TRANSIENT_FAILURE)) {
+        // each subchannel is responsible for its own backoff
+        return;
+      } else if (newState == IDLE) {
+        // coming out of backoff
+        subchannel.requestConnection();
+        return;
+      }
+    }
+
+    // update subchannel state unless it was in transient failure
+    if (states.get(subchannel) != TRANSIENT_FAILURE || newState != CONNECTING) {
+      states.put(subchannel, newState);
+    }
+
+    SubchannelPicker picker;
+    switch (newState) {
+      case IDLE:
+        // Subchannel was shutdown when ready: connect from beginning when prompted
+        addressIndex.reset();
+        picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+        updateBalancingState(IDLE, picker);
+        break;
+      case CONNECTING:
+        // It's safe to use RequestConnectionPicker here, so when coming from IDLE we could leave
+        // the current picker in-place. But ignoring the potential optimization is simpler.
+        if (currentState != TRANSIENT_FAILURE) {
+          picker = new Picker(PickResult.withNoResult());
+          updateBalancingState(CONNECTING, picker);
+        }
+        break;
+      case READY:
+        picker = new Picker(PickResult.withSubchannel(subchannel));
+        updateBalancingState(READY, picker);
+        shutdownRemaining(subchannel.getAddresses().getAddresses().get(0), subchannel);
+        break;
+      case TRANSIENT_FAILURE:
+        // If we are looking at current channel ...
+        if (subchannels.get(addressIndex.getCurrentAddress()).equals(subchannel)) {
+          // ... and it is the last address, report TRANSIENT_FAILURE.
+          // Sticky TRANSIENT_FAILURE will only allow this to happen once per address update.
+          addressIndex.increment();
+          if (enterTransientFailure()) {
+            iterate = false;
+            addressIndex.reset();
```

Comment:
I think we discussed this already. When address update to [3,2] the currentIndex would be reset to 0 pointing to [3]. Then when 2 failed, since it is not the currentIndex we are looking at, we ignore it. When 3 fails, we try to increment index pointing to [2]. When 2 reports failure, we will do helper.updateBalancingState to report TF.

The current implementation might partially work, but is still not I meant, there is still way to simplify, what about:
```
case TRANSIENT_FAULRE:
  if addressIndex invalid, no nothing; // end of iteration
  else if the subchannel corresponding to the current addresssIndex is not the subchannel in the callback, do nothing;
  else 
      increment addressIndex;
      requestConnection;
      if addressIndex is invalid, report TRANSIENT_FAILURE. // this means first time iteration done, do not reset index, keep the index at invalid.
  
```